\hypertarget{vcf2gtp_8py}{\subsection{vcf2gtp.\+py File Reference}
\label{vcf2gtp_8py}\index{vcf2gtp.\+py@{vcf2gtp.\+py}}
}


Convert .vcf files into .gtp format.  


\subsubsection*{Data Structures}
\begin{DoxyCompactItemize}
\item 
class \hyperlink{classvcf2gtp_1_1_map}{vcf2gtp.\+Map}
\begin{DoxyCompactList}\small\item\em Represents the genetic map. \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsubsection*{Functions}
\begin{DoxyCompactItemize}
\item 
\hypertarget{namespacevcf2gtp_aca110f564943bb4cb5347fd456c5d612}{def {\bfseries vcf2gtp.\+usage}}\label{namespacevcf2gtp_aca110f564943bb4cb5347fd456c5d612}

\item 
def {\bfseries vcf2gtp.\+V\+C\+Fparse}
\end{DoxyCompactItemize}
\subsubsection*{Variables}
\begin{DoxyCompactItemize}
\item 
\hypertarget{namespacevcf2gtp_a51af4083420d5d86917c4115ecfcfaed}{{\bfseries vcf2gtp.\+map\+Fname} = None}\label{namespacevcf2gtp_a51af4083420d5d86917c4115ecfcfaed}

\item 
\hypertarget{namespacevcf2gtp_a1ed17ca75c2b59c8a056354b9fef57e5}{{\bfseries vcf2gtp.\+map} = None}\label{namespacevcf2gtp_a1ed17ca75c2b59c8a056354b9fef57e5}

\item 
\hypertarget{namespacevcf2gtp_ab05e65af3ec334e11ceb0e77b54993a8}{{\bfseries vcf2gtp.\+vcf\+File} = None}\label{namespacevcf2gtp_ab05e65af3ec334e11ceb0e77b54993a8}

\item 
\hypertarget{namespacevcf2gtp_afa481fba4eb4976d40f9b755632234dd}{float {\bfseries vcf2gtp.\+recomb} = 0.\+0}\label{namespacevcf2gtp_afa481fba4eb4976d40f9b755632234dd}

\item 
\hypertarget{namespacevcf2gtp_aedf1c0fa60b9602ae85e8b5925ba21f1}{{\bfseries vcf2gtp.\+extrapolate} = False}\label{namespacevcf2gtp_aedf1c0fa60b9602ae85e8b5925ba21f1}

\item 
\hypertarget{namespacevcf2gtp_a9719632b74102a9b80d73557e5a2006c}{int {\bfseries vcf2gtp.\+i} = 1}\label{namespacevcf2gtp_a9719632b74102a9b80d73557e5a2006c}

\item 
\hypertarget{namespacevcf2gtp_aa560dce4c7d5fa35c33bdd80f1082cc1}{list {\bfseries vcf2gtp.\+vcf\+Fname} = sys.\+argv\mbox{[}i\mbox{]}}\label{namespacevcf2gtp_aa560dce4c7d5fa35c33bdd80f1082cc1}

\item 
\hypertarget{namespacevcf2gtp_a6df29e970b4b7c26ee4582da0cd38d09}{tuple {\bfseries vcf2gtp.\+line} = vcf\+File.\+readline()}\label{namespacevcf2gtp_a6df29e970b4b7c26ee4582da0cd38d09}

\item 
\hypertarget{namespacevcf2gtp_a3314b5a2eab02b79f3d582de8e74ccd4}{int {\bfseries vcf2gtp.\+snpid} = 0}\label{namespacevcf2gtp_a3314b5a2eab02b79f3d582de8e74ccd4}

\item 
\hypertarget{namespacevcf2gtp_aeaa5ecb08738c0b6631e02d10554fcf8}{tuple {\bfseries vcf2gtp.\+mappos} = map.\+get\+Map\+Pos(nucpos0)}\label{namespacevcf2gtp_aeaa5ecb08738c0b6631e02d10554fcf8}

\item 
\hypertarget{namespacevcf2gtp_a84f0a3d751c8d61e5212c2f54582ba9b}{string {\bfseries vcf2gtp.\+msg} = \char`\"{}E\+R\+R\+: vcf file contains multiple chromosomes\+: \%s, \%s.\textbackslash{}n\char`\"{}}\label{namespacevcf2gtp_a84f0a3d751c8d61e5212c2f54582ba9b}

\item 
\hypertarget{namespacevcf2gtp_aca0c13cab980a6d5e0d9c57ded41eb99}{{\bfseries vcf2gtp.\+nucpos0} = nucpos}\label{namespacevcf2gtp_aca0c13cab980a6d5e0d9c57ded41eb99}

\end{DoxyCompactItemize}


\subsubsection{Detailed Description}
Convert .vcf files into .gtp format. 

\label{vcf2gtp_8py_vcf2gtp}%
\hypertarget{vcf2gtp_8py_vcf2gtp}{}%
\subsubsection*{\hyperlink{vcf2gtp_8py}{vcf2gtp.\+py}, a program that converts .vcf files into .gtp format }

\begin{DoxyVerb}usage: vcf2gtp [options filename.vcf geneticMapFile]
where options may include:

  -r <x> or --recombPerBP <x>     Set recombination rate per nucleotide.

If two filenames are provided, the first must be a vcf file and the
second a file of recombination rates. In this case, the -r option is
not allowed.

If one filename is provided, it must be a vcf file, and the -r option
is mandatory.

If no filename is provided, the program reads vcf data from standard
input and the -r option is mandatory.

The program writes to standard output.
\end{DoxyVerb}


If no file is provided for recombination rates, the rate must be specified using the {\ttfamily -\/r} or {\ttfamily -\/-\/recomb\+Per\+B\+P} option, and the program generates map distances assuming that the recombination rate between two sites is proportional to the number of nucleotides that separate them.

The genetic map file must have a format like this\+: \begin{DoxyVerb}position COMBINED_rate(cM/Mb) Genetic_Map(cM)
14431347 9.6640973708 0
14432618 9.7078062447 0.0122830678
14433624 9.7138922111 0.0220491208
14433659 9.7163435060 0.0223891071
14433758 9.7078087850 0.0233510251
\end{DoxyVerb}


The first line is a list of column labels. After that, each row corresponds to a \hyperlink{struct_s_n_p}{S\+N\+P}. The first column gives the position, in nucleotides from the end of the chromosome. The second column gives the recombination rate, in c\+M per Mb, between that \hyperlink{struct_s_n_p}{S\+N\+P} and the next one, and the third column gives the map distance in c\+M. This is the format used by the 1000-\/\+Genomes Project

\begin{DoxyAuthor}{Author}
Ryan Bohlender and Alan Rogers 
\end{DoxyAuthor}
\begin{DoxyCopyright}{Copyright}
Copyright (c) 2014, Ryan Bohlender and Alan R. Rogers \href{mailto:rogers@anthro.utah.edu}{\tt rogers@anthro.\+utah.\+edu}. This file is released under the Internet Systems Consortium License, which can be found in file \char`\"{}\+L\+I\+C\+E\+N\+S\+E\char`\"{}. 
\end{DoxyCopyright}
